File: backend/app/schemas/review.py
"""Reviewer-only request and response contracts."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator

from backend.app.orchestrator.constants import WorkflowStatus
from backend.app.schemas.domain import (
    ClaimContext,
    FraudAssessmentContext,
    HumanDecision,
    HumanDecisionContext,
    PolicyContext,
)
from backend.app.schemas.orchestrator import AuditEvent


class _ReviewContract(BaseModel):
    model_config = ConfigDict(extra="forbid")


class HumanDecisionRequest(_ReviewContract):
    decision: HumanDecision
    reason: str = Field(min_length=3, max_length=1000)
    notes: str | None = Field(default=None, max_length=3000)

    @field_validator("reason")
    @classmethod
    def validate_reason(cls, value: str) -> str:
        value = value.strip()
        if len(value) < 3:
            raise ValueError("reason must contain at least 3 characters")
        return value

    @field_validator("notes")
    @classmethod
    def normalize_notes(cls, value: str | None) -> str | None:
        if value is None:
            return None
        value = value.strip()
        return value or None


class ReviewQueueItem(_ReviewContract):
    workflow_id: str
    claim_id: str
    created_at: datetime
    updated_at: datetime
    claim_status: str | None = None
    incident_type: str | None = None
    incident_date: date | None = None
    location: str | None = None
    risk_level: str | None = None
    risk_score: float | None = Field(default=None, ge=0, le=1)
    recommended_action: str | None = None
    risk_indicator_count: int = 0
    missing_documents: list[str] = Field(default_factory=list)


class ReviewQueueResponse(_ReviewContract):
    items: list[ReviewQueueItem]
    limit: int
    offset: int
    returned: int


class ReviewDetailResponse(_ReviewContract):
    workflow_id: str
    status: WorkflowStatus
    claim: ClaimContext
    policy: PolicyContext | None = None
    retrieval_context: dict[str, Any] | None = None
    fraud_assessment: FraudAssessmentContext
    audit_timeline: list[AuditEvent] = Field(default_factory=list)
    human_decision: HumanDecisionContext | None = None


class HumanDecisionResponse(_ReviewContract):
    workflow_id: str
    claim_id: str
    status: WorkflowStatus
    decision: HumanDecisionContext
    message: str

File: frontend/src/types/review.ts
export type RiskLevel = "low" | "medium" | "high" | "unassessed";

export interface ReviewerSummary {
  claim_overview: string;
  policy_findings: string[];
  risk_observations: string[];
  missing_items: string[];
  reviewer_action_points: string[];
}

export interface ReviewQueueItem {
  workflow_id: string;
  claim_id: string;
  created_at: string;
  updated_at: string;
  claim_status: string | null;
  incident_type: string | null;
  incident_date: string | null;
  location: string | null;
  risk_level: RiskLevel | null;
  risk_score: number | null;
  recommended_action: string | null;
  risk_indicator_count: number;
  status?: string;
  assigned_to?: string | null;
  assigned_at?: string | null;
  missing_documents?: string[];
  reviewer_summary?: ReviewerSummary | null;
}

export interface ReviewQueueResponse {
  items: ReviewQueueItem[];
  limit: number;
  offset: number;
  returned: number;
}

export interface ClaimAssignmentRequest {
  assigned_to: string;
}

export interface ClaimAssignmentResponse {
  assignment_id: string;
  workflow_id: string;
  claim_id: string;
  assigned_to: string;
  assigned_by: string;
  assigned_at: string;
  status: string;
  message: string;
}

export interface HumanDecisionRequest {
  decision: "approve" | "reject" | "request_more_information" | "escalate";
  reason: string;
  notes?: string | null;
  settlement_amount?: number | null;
}

export interface HumanDecisionResponse {
  workflow_id: string;
  claim_id: string;
  decision_id: string;
  decision: string;
  claim_status: string;
  decided_at: string;
  status: string;
  message: string;
}

export interface ClaimDocumentItem {
  document_id: string;
  claim_id: string;
  customer_id?: string;
  document_type: string;
  original_filename: string;
  file_size_bytes?: number;
  content_type?: string;
  uploaded_at?: string;
}

export interface ClaimContextDetails {
  claim_id?: string;
  claim_reference?: string;
  customer_id?: string;
  incident_type?: string;
  incident_date?: string;
  incident_location?: string;
  incident_description?: string;
  claimed_amount?: number | null;
  claim_status?: string;
  policy_id?: string;
  policy_number?: string;
}

export interface PolicyContextDetails {
  policy_id?: string;
  policy_number?: string;
  customer_id?: string;
  status?: string;
  start_date?: string;
  end_date?: string;
  coverage_details?: Record<string, unknown>;
  exclusions?: string[];
}

export interface FraudIndicatorDetails {
  rule_name?: string;
  description?: string;
  severity?: string;
}

export interface FraudAssessmentDetails {
  risk_level?: string;
  risk_score?: number | null;
  recommended_action?: string;
  indicators?: FraudIndicatorDetails[];
}

export interface ReviewerGuidanceResultDetails {
  status?: string;
  data?: {
    reviewer_summary?: ReviewerSummary;
  };
}

export interface HumanDecisionDetails {
  decision_id?: string;
  workflow_id?: string;
  claim_id?: string;
  decision?: string;
  reviewer_id?: string;
  reviewer_role?: string;
  reason?: string;
  notes?: string;
  decided_at?: string;
}

export interface ReviewDetailResponse {
  workflow_id: string;
  status: string;
  claim: ClaimContextDetails;
  policy?: PolicyContextDetails;
  retrieval_context?: Record<string, unknown>;
  fraud_assessment?: FraudAssessmentDetails;
  reviewer_guidance_result?: ReviewerGuidanceResultDetails;
  audit_timeline?: Record<string, unknown>[];
  human_decision?: HumanDecisionDetails;
  documents: ClaimDocumentItem[];
  assigned_to?: string | null;
  assigned_at?: string | null;
}
